dev/src/reference.ts

```diff
@@ -2088,21 +2205,27 @@ export class Query<T = firestore.DocumentData> implements firestore.Query<T> {
       transform: (proto, enc, callback) => {
         const readTime = Timestamp.fromProto(proto.readTime);
         if (proto.document) {
-          const document = this.firestore.snapshot_(
-            proto.document,
-            proto.readTime
-          );
-          const finalDoc = new DocumentSnapshotBuilder<T>(
-            document.ref.withConverter(this._queryOptions.converter)
-          );
-          // Recreate the QueryDocumentSnapshot with the DocumentReference
-          // containing the original converter.
-          finalDoc.fieldsProto = document._fieldsProto;
-          finalDoc.readTime = document.readTime;
-          finalDoc.createTime = document.createTime;
-          finalDoc.updateTime = document.updateTime;
-          lastReceivedDocument = finalDoc.build() as QueryDocumentSnapshot<T>;
-          callback(undefined, {document: lastReceivedDocument, readTime});
+          if (this._queryOptions.ancestorType) {
+            const path = ResourcePath.EMPTY.append(proto.document.name);
+            const docRef = new DocumentReference(this._firestore, path);
+            callback(undefined, {document: docRef});
+          } else {
+            const document = this.firestore.snapshot_(
+              proto.document,
+              proto.readTime
+            );
+            const finalDoc = new DocumentSnapshotBuilder<T>(
+              document.ref.withConverter(this._queryOptions.converter)
+            );
+            // Recreate the QueryDocumentSnapshot with the DocumentReference
+            // containing the original converter.
+            finalDoc.fieldsProto = document._fieldsProto;
+            finalDoc.readTime = document.readTime;
+            finalDoc.createTime = document.createTime;
+            finalDoc.updateTime = document.updateTime;
+            lastReceivedDocument = finalDoc.build() as QueryDocumentSnapshot<T>;
```

Comment:
You probably want to hit the code path because `lastReceivedDocument` implements implicit query resumption for deadline failures.

Reply:
Done.